### PR TITLE
feat: scheduler (12/): add more scheduler logic 

### PR DESCRIPTION
**Pull request by michaelawyu**
### Description of your changes

This PR is part of the PRs that implement the Fleet workload scheduling.

It features more scheduling logic for PickAll type CRPs.

I have:

- [x] Run `make reviewable` to ensure this PR is ready for review.

### How has this code been tested

- [x] Unit tests

### Special notes for your reviewer

To control the size of the PR, certain unit tests are not checked in; they will be sent in a separate PR.

### CODE REVIEW

**Comment by michaelawyu** [comment]
A special note:

Currently the API has a limit of 100 for cluster decisions, but it also dictates that we need to keep all decisions from selected clusters. Considering that we do not have a schema-side range set for numOfClusters, do we drop decisions when there are over 100 selected clusters or do we set a limit of 100 to numOfClusters as well?

**Comment by ryanzhang-oss** [comment]
> A special note:
> 
> Currently the API has a limit of 100 for cluster decisions, but it also dictates that we need to keep all decisions from selected clusters. Considering that we do not have a schema-side range set for numOfClusters, do we drop decisions when there are over 100 selected clusters or do we set a limit of 100 to numOfClusters as well?

I wonder where is the 100 limit come from? I think k8s scheduler has a parameter about how many nodes to consider. Do we need to add that?

**Comment by zhiying-lin** [comment]
> A special note:
> 
> Currently the API has a limit of 100 for cluster decisions, but it also dictates that we need to keep all decisions from selected clusters. Considering that we do not have a schema-side range set for numOfClusters, do we drop decisions when there are over 100 selected clusters or do we set a limit of 100 to numOfClusters as well?

we cannot limit the numberOfClusters for the selectAll type and still need to handle the 100 limit size.

> > A special note:
> > Currently the API has a limit of 100 for cluster decisions, but it also dictates that we need to keep all decisions from selected clusters. Considering that we do not have a schema-side range set for numOfClusters, do we drop decisions when there are over 100 selected clusters or do we set a limit of 100 to numOfClusters as well?
> 
> I wonder where is the 100 limit come from? I think k8s scheduler has a parameter about how many nodes to consider. Do we need to add that?

100 limit comes from https://github.com/Azure/fleet/blob/main/apis/placement/v1beta1/policysnapshot_types.go#L77

File: pkg/scheduler/framework/framework.go [hunk 1]
```diff
@@ -354,7 +358,16 @@ func (f *framework) runSchedulingCycleForPickAllPlacementType(
 		return ctrl.Result{}, err
```
Comment (zhiying-lin) [inline pkg/scheduler/framework/framework.go:358]: missed one comment before, once we create the bindings, should we add the finalizer in the crp to make sure all of the bindings are deleted by scheduler before we delete the crp?

File: pkg/scheduler/framework/frameworkutils.go [hunk 2]
```diff
@@ -183,3 +184,67 @@ func crossReferencePickedCustersAndObsoleteBindings(
 
 	return toCreate, toUpdate, toDelete, nil
 }
+
+// newSchedulingDecisionsFrom returns a list of scheduling decisions, based on the newly manipulated list of
+// bindings and (if applicable) a list of filtered clusters.
+func newSchedulingDecisionsFrom(maxClusterDecisionCount int, filtered []*filteredClusterWithStatus, existing ...[]*fleetv1beta1.ClusterResourceBinding) []fleetv1beta1.ClusterDecision {
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply pkg/scheduler/framework/frameworkutils.go:190]: The purpose of `maxClusterDecisionCount` is not for limiting the number of total decisions, as its comment has explained. Its purpose, instead, is to make sure that we add the right number of decisions about clusters that are not picked to the status

File: pkg/scheduler/framework/frameworkutils.go [hunk 3]
```diff
@@ -183,3 +184,67 @@ func crossReferencePickedCustersAndObsoleteBindings(
...
+	// Pre-allocate with a reasonable capacity.
+	newDecisions := make([]fleetv1beta1.ClusterDecision, 0, maxClusterDecisionCount)
+
+	// Build new scheduling decisions.
+	for _, bindingSet := range existing {
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply pkg/scheduler/framework/frameworkutils.go:195]: That's why I added the special note in the PR. Adding a limit won't do here as it also violates our specification that all picked clusters will be included in the status.

File: pkg/scheduler/framework/frameworkutils.go [hunk 4]
```diff
@@ -183,3 +184,67 @@ func crossReferencePickedCustersAndObsoleteBindings(
...
+		desiredDecisionByCluster[decision.ClusterName] = decision
+	}
+
+	for _, decision := range current {
+		matched, ok := desiredDecisionByCluster[decision.ClusterName]
+		if !ok {
+			// No matching decision can be found.
+			return false
+		}
+		if !reflect.DeepEqual(decision, matched) {
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply pkg/scheduler/framework/frameworkutils.go:243]: Ah, actually a lot of Kubernetes code uses [reflect.DeepEqual](https://github.com/search?q=repo%3Akubernetes%2Fkubernetes%20reflect.DeepEqual&type=code). And the Semantic package is just wrapper around reflect.Equalities; considering that the decision object does not have any field that requires semantics based equalities provided by the Semantic package, reflect.DeepEqual should suffice here.

**Comment by michaelawyu** [comment]
> > A special note:
> > Currently the API has a limit of 100 for cluster decisions, but it also dictates that we need to keep all decisions from selected clusters. Considering that we do not have a schema-side range set for numOfClusters, do we drop decisions when there are over 100 selected clusters or do we set a limit of 100 to numOfClusters as well?
> 
> I wonder where is the 100 limit come from? I think k8s scheduler has a parameter about how many nodes to consider. Do we need to add that?

Hi Ryan! The API has a maxItems = 100 limit on ClusterDecisions, so it may be in conflict with the status updating process if the customer picks over 100 clusters. This does not concern the node threshold much I think; the threshold does not kick unless the cluster has over 100 nodes IIRC. I guess it's kind of safe right now for us to assume that most fleets will not have over 100 member clusters (the design target was 500?)

File: pkg/scheduler/framework/frameworkutils.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (michaelawyu) [reply pkg/scheduler/framework/frameworkutils.go:197]: Discussed offline.

The API side limit has been raised to 1000; we will cap decisions to this limit. The behavior for adding unselected cluster decisions remains the same.

File: pkg/scheduler/framework/frameworkutils.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (michaelawyu) [reply pkg/scheduler/framework/frameworkutils.go]: Discussed offline.

The API side limit has been raised to 1000; we will cap decisions to this limit. The behavior for adding unselected cluster decisions remains the same -> the 20 value is user-configurable via CLI; we will not provide reasons why a cluster is not selected if there are enough decisions already for selected clusters..

File: pkg/scheduler/framework/framework.go (hunk 1 above)
Comment (michaelawyu) [reply pkg/scheduler/framework/framework.go:364]: Oh, that one is added at the upper layer by the scheduler, not the scheduler framework.

File: pkg/scheduler/framework/frameworkutils.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (zhiying-lin) [reply pkg/scheduler/framework/frameworkutils.go:251]: equality.Semantic.DeepEqual will have "// An empty slice *is* equal to a nil slice for our purposes; same for maps.".

Do we want it for the "ClusterScore *ClusterScore " in the "ClusterDecision"?

File: pkg/scheduler/framework/frameworkutils.go [hunk 5]
```diff
@@ -187,19 +188,27 @@ func crossReferencePickedCustersAndObsoleteBindings(
...
+		slotsLeft -= setLength
+		if slotsLeft <= 0 {
+			klog.V(2).InfoS("Reached API limit of cluster decision count; decisions off the limit will be discarded")
+			break
 		}
 	}
 
 	// Move some decisions from unbound clusters, if there are still enough room.
-	if diff := maxClusterDecisionCount - len(newDecisions); diff > 0 {
+	if diff := maxUnselectedClusterDecisionCount - len(newDecisions); diff > 0 {
```
Comment (zhiying-lin) [inline pkg/scheduler/framework/frameworkutils.go:211]: now the maxUnselectedClusterDecisionCount is renamed. should we update the logic here?

based on the comment, 

"// maxUnselectedClusterDecisionCount controls the maximum number of decisions for unselected clusters
	// added to the policy snapshot status."

 If i understand it correctly, we need to populate min(maxUnselectedClusterDecisionCount, slotsLeft) using filteredClusters?

File: pkg/scheduler/framework/frameworkutils.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (michaelawyu) [reply pkg/scheduler/framework/frameworkutils.go:251]: ClusterScore is neither a slice or a map?

File: pkg/scheduler/framework/frameworkutils.go (hunk 5 above)
Comment (michaelawyu) [reply pkg/scheduler/framework/frameworkutils.go:211]: Not really. As the comment has explained, the limit is added to make sure that we only populate the status with info. for filtered out clusters if the total number of decisions is small enough; i.e. we won't fill all the slots blindly with decisions for unfiltered clusters.

File: pkg/scheduler/framework/frameworkutils.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (zhiying-lin) [reply pkg/scheduler/framework/frameworkutils.go:251]: my bad, somehow i treated it as the pointer of the array.

File: pkg/scheduler/framework/frameworkutils.go (hunk 5 above)
Comment (zhiying-lin) [reply pkg/scheduler/framework/frameworkutils.go:211]: i feel the original field name is better. I was confused by the new name and its meaning. :)

**Review by zhiying-lin — DISMISSED** [review]
LGTM
nit, i feel maxClusterDecisionCount is better than the current one. leaving the decision to you :)

**Comment by michaelawyu** [comment]
I will keep the old name for now and see if there's a better name.

**Pull request merged**